Add wxSystemAppearance to check for dark mode under macOS

vadz opened the pull request:
Provide a way to retrieve the name of the current system appearance
(mostly for diagnostic purposes) and check if it uses predominantly dark
colours.

Currently this class has a non-trivial (but still very simple)
implementation under macOS only, but it could be extended to other
platforms later.

---

Stefan, this is, finally, as trivial as that, but I'd still like you to look over it and let me know if I did anything stupid. I'd also like to ask if you can advise for the best way to get rid of the warnings described in this [wx-dev post](https://groups.google.com/d/msg/wx-dev/_uOvYJzkI3Q/D4WOQO_oBwAJ).

Any suggestions for the API changes/extensions are welcome too, before I document it: as you can see I kept just `IsDark()` instead of more general `GetType()` returning `DARK`, `LIGHT` or `UNKNOWN` because I couldn't really think of a situation in which the latter would be actually useful, but if you can, please let me know, it's not too late to change it yet. TIA!

vadz committed:
Add wxSystemAppearance to check for dark mode under macOS

Provide a way to retrieve the name of the current system appearance
(mostly for diagnostic purposes) and check if it uses predominantly dark
colours.

Currently this class has a non-trivial (but still very simple)
implementation under macOS only and simply checks whether the default
text colour is brighter than the default background colour under the
other platforms, but other platform-specific implementations could be
added later.

Also update the drawing sample "system colours" page to show the system
appearance as well.

csomor commented:
It's ok, although I still think, that `IsDark()` or another appearance method should compare some system color settings, not tying things to names, and there is a dark mode in win 10 as well, and it too could be determined that way ...

vadz commented:
We could have `UsesDarkBackground()`, but explaining the difference between it and `IsDark()` is going to be fun. We'd need to rename the latter to something like `IsStandardDarkMode()` and compare its name instead of using "best match" then, probably.

Would you prefer this?

csomor commented:
We could leave IsDark() the way it is now for macOS, and perhaps have `GetBackgroundBrightness` or something like that averaging the RGB of `wxSYS_COLOUR_WINDOW`. Now on Windows we could use this method to implement `IsDark()` ?

vadz commented:
As per [this message](https://groups.google.com/d/msg/wx-dev/1vXBqmsHBEw/5Zj6xkPyBgAJ), there actually seems to be a better way to do this under MSW too.

So my final proposal is this: let's keep this implementation of `IsDark()` but add (in another PR) `GetBackgroundBrightness()` and implement `IsDark()` by default in its terms, i.e. use it under pre-10.14 macOS as well as all the other platforms (with the possible exception of Windows if the registry key mentioned in the link above exists). What do you think?

csomor commented:
This sounds perfect to me :-) Thanks

vadz committed:
Add wxColour::GetLuminance()

This method can be used to return the perceived brightness of the
colour.

vadz committed:
Use wxSystemAppearance::IsDark() to check for dark mode in wxOSX

This should be more reliable than simply checking the red component of
the default window background colour.

vadz commented:
Here is the updated version. It also includes a commit from #1300 which is needed for `IsUsingDarkBackground()` implementation, please ignore it when reviewing this one.

What do you think about it? Also pinging @NewPagodi.

NewPagodi commented:
This looks good to me.

csomor reviewed (approved):
Looks good to me, and although as I said I would prefer rendering code to just compare whether it's a dark background however it's named, I'm fine with this. Thanks.
